Add Heroku support with `rake assets:precompile` hook

mattbrictson opened the pull request:
This PR adds automatic (almost) Heroku support to Torba.

It works by adding a Rake task to Rails apps called `torba:pack`. This task is installed as a *prerequisite* for `assets:precompile`, ensuring that it is automatically run by the standard Heroku Ruby buildpack, as [discussed here](https://github.com/heroku/heroku-buildpack-ruby/issues/442#issuecomment-164174721).

The only manual step is setting Heroku-friendly `TORBA_HOME_PATH` and `TORBA_CACHE_PATH` config vars. This ensures that Torba's cache is shared across deploys. I documented this in the README.

I tested this with a brand new Rails 4.2.5 app deployed to Heroku, and it works great.

**Caveat:** I had to change the behavior of Torba's verification. For Rails specifically, I added a `torba/rails/verify.rb` that differs from the standard `torba/verify.rb` in that it disables verification when Rake is loaded. I had to do this because it was impossible to run `rake assets:precompile` otherwise.

@nashbridges Does this look good to you? If so I will clean up the commits and work on adding tests.

Fixes #4.

nashbridges commented:
Awesome!

Do you think that we need separate "rails/verify"? For example, wouldn't https://github.com/kalasjocke/sinatra-asset-pipeline suffer from the same issue?

nashbridges commented on README.md:
```diff
@@ -251,10 +251,43 @@ default), except that Torba does have a way to specify exact list of files to im
+`TORBA_HOME_PATH` should be within your app so that Torba's assets are included
+in the slug.
+
+`TORBA_TEMP_PATH` should make use of the `tmp/cache/assets` directory. This is
```
TORBA_CACHE_PATH

mattbrictson commented:
OK, I have fixed/added the following and this PR is now ready for final review.

* Fixed typo in README
* Removed `rails/verify.rb` and moved its logic to `verify.rb`
* Added unit tests
* Added `Torba::RakeTask` documentation
* Consolidated into 3 commits

nashbridges commented on lib/torba/verify.rb:
```diff
@@ -1,7 +1,14 @@
+# this case it is OK that the Torbafile hasn't been packed yet, because we are
+# about to do so.
+#
+unless defined?(Rake::Task)
```
Unfortunately, this doesn't work if I run `bin/rake assets:precompile` or `bundle exec rake assets:precompile`.
Could we also/instead check if `$0` ends with "rake" to cover migration/compilation tasks?
Keep in mind that in Rails 5 one [can use](https://github.com/rails/rails/commit/3e65c3d3886336e9145438cdeacaf4ebec6a48b8) `bin/rails assets:precompile`

mattbrictson replied:
Given that `bin/rails` can be used, then we can't rely on `$0`. And if `bin/rake` or `bin/rails` binstubs are used, then Rails is loaded *before* Rake is loaded, so we can't rely on `defined?(Rake)`.

The only thing I can think of is checking `ARGV.include?("assets:precompile")`. That would mean the logic becomes:

```ruby
unless defined?(Rake::Task) || ARGV.include?("assets:precompile")
  # verify ...
end
```

Should I go with that?

nashbridges replied:
The results of my experiments:

rails 4.2 (w/o spring):

command | defined?(Rake::Task) | $0
------------- | ---------------------------- | -------
bin/rake db:migrate | nil | "bin/rake"
bin/rake assets:precompile | nil | "bin/rake"
bundle exec rake db:migrate | "constant" | "/Users/nash/.rbenv/versions/2.2/bin/rake"
bundle exec rake assets:precompile | "constant" | "/Users/nash/.rbenv/versions/2.2/bin/rake"

rails 5.0.0.beta.1.1 (w/ srping)

command | defined?(Rake::Task) | $0
------------- | ---------------------------- | -------
bin/rake db:migrate | nil | "spring app    \| rails5 \| started 0 secs ago \| development mode"
bin/rake assets:precompile | nil | "spring app    \| rails5 \| started 0 secs ago \| development mode"
bin/rails db:migrate | nil | "bin/rails"
bin/rails assets:precompile | nil | "bin/rails"
bundle exec rake db:migrate | "constant" | "/Users/nash/.rbenv/versions/2.2/bin/rake"
bundle exec rake assets:precompile | "constant" | "/Users/nash/.rbenv/versions/2.2/bin/rake"

I assume 5x w/o spring is the same as 4.2.

You're correct, we can't rely neither on `Rake::Task`, nor on `$0`. Maybe use `ARGV`, but use whitelist approach instead? We only need assets related verification when "server" is run or `RAILS_ENV` is test.

nashbridges replied:
Or maybe we should give up and leave your check + test for env variable like `TORBA_DONT_VERIFY` and then users will append it to sensitive commands.

nashbridges replied:
BTW, if we treat this as a general approach, then `Rake` constant could also be not defined.

mattbrictson replied:
I agree. It looks like we are going in circles and there is no good way to auto-detect. I'll use `TORBA_DONT_VERIFY` instead.

mattbrictson commented:
@nashbridges I added the `TORBA_DONT_VERIFY` environment variable that we were discussing. Is this now a good enough solution?

nashbridges commented:
Merged and [released](https://rubygems.org/gems/torba/versions/0.5.0).
Thanks for the great work!
